### PR TITLE
feat(moe): enable MxInt4 in unified API

### PR DESCRIPTION
**Pull request by feih-nv**
## 📌 Description

Enable TensorRT-LLM MxInt4 execution through FlashInfer’s unified MoE API.

This change adds production BF16-to-MxInt4 weight preparation through `TrtllmMxInt4Config.prepare_weights()`, including signed INT4 packing, BF16 block scales, gated-activation row permutation, scale interleaving, and BlockMajorK conversion.

It also adds and registers `TrtllmMxInt4RoutedRunner`, supporting:
- Packed precomputed routing
- MxInt4 `FromLogits` support with BF16 router logits
- Expert-parallel local expert offsets
- Unified MoELayer dispatch and autotuning
- CUDA graph capture and replay

The unified MoE fuzzer now includes an MxInt4 handler, a quantization-aware reference, and curated packed and `FromLogits` seeds.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see the [pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All relevant tests are passing.

### SM100

- Unified MxInt4 suite: 30 passed
- Curated unified MoE fuzzer seeds: 2 passed
- Existing flat routed regressions: 4 passed

### SM107

- Unified MxInt4 suite: 30 passed
- Curated unified MoE fuzzer seeds: 2 passed

## Summary by CodeRabbit

* **New Features**
  * Added TRTLLM MxInt4 support for unified Mixture-of-Experts layers.
  * Added BF16 activation and packed INT4 weight preparation for supported routed architectures.
  * Added precomputed and in-kernel routing, autotuning, and CUDA graph compatibility.
  * Added validation for supported data types, shapes, devices, and weight layouts.

* **Bug Fixes**
  * Improved permutation cache accuracy by accounting for all relevant configuration settings.

* **Tests**
  * Expanded fuzz and integration coverage for MxInt4 quantization, routing, validation, and execution.

### CODE REVIEW

**Comment by feih-nv** [comment]
@coderabbitai review

File: flashinfer/fused_moe/prepare.py [hunk 1]
```diff
@@ -971,6 +972,137 @@ def prepare_trtllm_fp8_per_tensor_activations(
...
+            permute_cache, w2_q[expert], epilogue_tile_m
+        )
+        # Keep the established flat-test MxInt4 scale permutation contract;
+        # preparation parity tests cover this asymmetric GEMM1/GEMM2 setting.
+        w2_scale_permute = get_w2_permute_indices_with_cache(
+            permute_cache,
+            w2_sf[expert],
+            epilogue_tile_m,
+            num_elts_per_sf=16,
+        )
```
[the remark this answers is not in the document]
Comment (feih-nv) [reply flashinfer/fused_moe/prepare.py:1076]: this should be addressed by commit c4415b286c1bac30a91c9d08a9b3d39d18418a29

[the remark this answers is not in the document]
Comment (feih-nv) [reply flashinfer/fused_moe/prepare.py:1076]: @coderabbitai the cache keys now include `num_elts_per_sf`, `epilogue_tile_m`, and `is_gated_act_gemm`, so separate cache dictionaries are not required.

File: flashinfer/fused_moe/core.py [hunk 2]
```diff
@@ -170,8 +170,14 @@ def _maybe_get_cached_w3_w1_permute_indices(
...
-    # Create a unique cache key (weight_type, weight_shape)
-    cache_key = ("w3_w1", dst_w3_w1_weight.shape)
+    # Include every parameter that changes the generated permutation.
+    cache_key = (
+        "w3_w1",
+        dst_w3_w1_weight.shape,
+        epilogue_tile_m,
+        num_elts_per_sf,
+        is_gated_act_gemm,
+    )
```
[the remark this answers is not in the document]
Comment (feih-nv) [reply flashinfer/fused_moe/core.py:180]: All current production cache consumers move the returned permutation indices to the target tensor’s device before indexing (p.to(device) / permute.to(device)), so cross-device cache reuse is correct. Device-scoped keys could avoid an extra cross-device copy, but that would be a performance optimization rather than a correctness fix.

**Review by aleozlx — APPROVED** [review]
